### PR TITLE
Pluggable project model

### PR DESCRIPTION
**Pull request by matklad**
This is a WIP at abstracting the project model, as proposed in https://github.com/racer-rust/racer/pull/855#issuecomment-392448281.

I hope that it'll solve https://github.com/rust-lang-nursery/rls/issues/956, but I haven't implemented the RLS side of things yet, so there's a chance that this whole approach breaks down.

Nevertheless, I figured I'd just toss it over the fence for now to see what the folks think :)

### CODE REVIEW

**Comment by kngwyu** [comment]
Great!
I feel we need to implement this kind of feature too, but now I don't have enough time, so really thank you for this PR.

But,  I think it's better to merge this PR after implementing `RacerProjectModelProvider` to RLS.

File: src/racer/project_model.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
+use std::path::{Path, PathBuf};
+
+pub trait RacerProjectModelProvider {
```
Comment (kngwyu) [inline src/racer/project_model.rs]: Do we really need `Racer` prefix here?
From other crates it looks `racer::RacerProjectModelProvider`

File: src/racer/project_model.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    impl<'c> RacerProjectModelProvider for CargoProjectModel<'c> {
+        fn discover_project_manifest(&self, path: &Path) -> Option<PathBuf> {
+            let path = cargo_try!(find_root_manifest_for_wd(path));
+            Some(path)
+        }
+        fn resolve_dependency(&self, manifest: &Path, libname: &str) -> Option<PathBuf> {
+            if self.get_deps(manifest).is_none() {
+                // cache doesn't exist
+                // calculating depedencies can be bottleneck we use info! here(kngwyu)
+                info!("[get_outer_crates] cache doesn't exist");
```
Comment (kngwyu) [inline src/racer/project_model.rs]: Could you please change function name here?

File: src/racer/project_model.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                let deps_map = self.resolve_dependencies(&manifest)?;
+                self.cache_deps(manifest, deps_map);
+            }
+            let deps = self.get_deps(manifest).unwrap();
+            debug!("[get_outer_crates] cache exists for manifest",);
```
Comment (kngwyu) [inline src/racer/project_model.rs]: Same as above, please change function name in debug!.

File: src/racer/project_model.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        fn resolve_dependency(&self, manifest: &Path, libname: &str) -> Option<PathBuf> {
+            if self.get_deps(manifest).is_none() {
+                // cache doesn't exist
+                // calculating depedencies can be bottleneck we use info! here(kngwyu)
+                info!("[get_outer_crates] cache doesn't exist");
+                let deps_map = self.resolve_dependencies(&manifest)?;
+                self.cache_deps(manifest, deps_map);
+            }
+            let deps = self.get_deps(manifest).unwrap();
```
Comment (kngwyu) [inline src/racer/project_model.rs]: Why call `self.get_deps` twice here

**Comment by kngwyu** [comment]
I left some comments, but for the whole code, it looks good to me.

**Comment by matklad** [comment]
The RLS sideof things is https://github.com/rust-lang-nursery/rls/issues/956. It's super ugly atm, but at least it works! Will look into clean-up of the racer implementation next! Thanks for the review.

**Comment by matklad** [comment]
Rebased and cleaned-up. [The RLS](https://github.com/rust-lang-nursery/rls/pull/965) side is super incomplete and ugly, but it works, and indeed get's rid of one Cargo :-)

**Comment by kngwyu** [comment]
Now looks good enough to merge :+1: 
But I'm going to wait a little, in case you come up with an alternative idea while working on RLS side :slightly_smiling_face:

**Pull request merged**